sharding/contracts/sharding_manager_test.go

```diff
@@ -1,130 +1,560 @@
+		t.Fatalf("Failed to deregister notary: %v", err)
 	}
-	if collatorStruct.Addr != addr {
-		t.Fatalf("Wrong collator address, %v should be %v", collatorStruct.Addr, addr)
+	backend.Commit()
+
+	numNotaries, _ = smc.NotaryPoolLength(&bind.CallOpts{})
+	if numNotaries.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 0, Got: %v", numNotaries)
 	}
-	if collatorStruct.Deposit.Cmp(collatorDeposit) != 0 {
-		t.Fatalf("Wrong collator deposit, %v should be %v", collatorStruct.Deposit, collatorDeposit)
+
+	// Notary 0 re-registers again
+	smc.RegisterNotary(txOpts)
+	backend.Commit()
+
+	numNotaries, _ = smc.NotaryPoolLength(&bind.CallOpts{})
+	if numNotaries.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 0, Got: %v", numNotaries)
 	}
+}
+
+func TestNotaryRelease(t *testing.T) {
+	addr := crypto.PubkeyToAddress(mainKey.PublicKey)
+	backend := backends.NewSimulatedBackend(core.GenesisAlloc{addr: {Balance: accountBalance2000Eth}})
+	txOpts := bind.NewKeyedTransactor(mainKey)
+	txOpts.Value = notaryDeposit
+	_, _, smc, _ := deploySMCContract(backend, mainKey)
 
-	// Check for the Deposit event
-	depositsEventsIterator, err := smc.FilterDeposit(&bind.FilterOpts{})
+	// Notary 0 registers
+	smc.RegisterNotary(txOpts)
+	backend.Commit()
+
+	notary, _ := smc.NotaryRegistry(&bind.CallOpts{}, addr)
+	numNotaries, _ := smc.NotaryPoolLength(&bind.CallOpts{})
+
+	if !notary.Deposited {
+		t.Fatalf("Notary has not registered. Got deposited flag: %v", notary.Deposited)
+	}
+
+	if numNotaries.Cmp(big.NewInt(1)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 1, Got: %v", numNotaries)
+	}
+
+	// Fast forward 10 blocks
+	for i := 0; i < 10; i++ {
+		backend.Commit()
+	}
+
+	// Notary 0 deregisters
+	txOpts = bind.NewKeyedTransactor(mainKey)
+	_, err := smc.DeregisterNotary(txOpts)
+	if err != nil {
+		t.Fatalf("Failed to deregister notary: %v", err)
+	}
+	backend.Commit()
+
+	numNotaries, _ = smc.NotaryPoolLength(&bind.CallOpts{})
+	if numNotaries.Cmp(big.NewInt(0)) != 0 {
+		t.Fatalf("Incorrect count from notary pool. Want: 0, Got: %v", numNotaries)
+	}
+
+	// Fast forward until lockup ends
+	for i := 0; i < int(sharding.NotaryLockupLength*sharding.PeriodLength+1); i++ {
+		backend.Commit()
+	}
+
+	// Notary 0 releases
+	_, err = smc.ReleaseNotary(txOpts)
 	if err != nil {
-		t.Fatalf("Failed to get Deposit event: %v", err)
+		t.Fatalf("Failed to release notary: %v", err)
 	}
-	if !depositsEventsIterator.Next() {
-		t.Fatal("No Deposit event found")
+	backend.Commit()
+
+	notary, err = smc.NotaryRegistry(&bind.CallOpts{}, addr)
+	if err != nil {
+		t.Fatalf("Can't get notary registry info: %v", err)
 	}
-	if depositsEventsIterator.Event.Collator != addr {
-		t.Fatalf("Collator address mismatch: %x should be %x", depositsEventsIterator.Event.Collator, addr)
+
+	if notary.Deposited {
+		t.Fatalf("Notary deposit flag should be false after released")
```

Comment:
I think this can be `t.Errorf(...)`. The general rule of go tests is to go as far as you can before calling Fatal. In this case, I think the test can proceed to the next part where the balance is checked. Please take a look at the test and see if there are other areas where you could call `t.Errorf` rather than `t.Fatal` because the test _could_ still continue (although would likely fail for the other conditions). It would be nice to see many errors, if there were many.